force-app/main/default/classes/DMLOperations.cls

```diff
@@ -139,17 +143,40 @@ public with sharing class DMLOperations {
      * Question 8
      * Insert or updates opportunities from list of names.
      * Associate the Opportunities with account that has the name accountName.
-     * There should be an Opportunity for each name provided in the list.
+     * There needs to be an Opportunity for each name provided in the list.
      * HINT: The account may or may not exists in the database but is required for opportunity creation. Consider how upsert operation can help with this.
      * @param accountName The name of the Account.
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        List <Account> Accounts = [SELECT Id FROM Account WHERE Name = :accountName]; // Query the account
+        Account varAccount = new Account(Name=accountName);
+        if(Accounts.isempty()){
+        insert varAccount;}
+        else {varAccount=[SELECT Id FROM Account WHERE Name = :accountName];}    
+        Map<String,Opportunity> opportunityNametoRecordMap = new Map<String,Opportunity>();
+        List <Opportunity> existingOpportunities = [SELECT Id, Name FROM Opportunity WHERE Name IN :oppNames];
+        for (Opportunity OppinList: existingOpportunities){
+            opportunityNametoRecordMap.put(OppinList.Name,OppinList); // Populate the map with the Name, Record
+        }
+        List <Opportunity> OppstoUpsert= new List <Opportunity>();
+        for (String oppName:OppNames){
+            Opportunity varOpp = new Opportunity();
+            if(OpportunityNametoRecordMap.keySet().contains(oppName)){//if the key set contains OppName
+                varOpp.AccountId=varAccount.Id;
+                varOpp.Id=opportunityNametoRecordMap.get(oppName).Id; // Define the Opportunity instance as the result of the Map's value, which is an Opp record, and provide the Id
+                varOpp.Name=oppName;
+                OppstoUpsert.add(varOpp);
+            }
+            else{
+                varOpp.AccountId=varAccount.Id;
+                varOpp.Name=oppName;
+                OppstoUpsert.add(varOpp);
+            }
+        }
+            upsert OppstoUpsert;
```

Comment:
Excellent job using the Map pattern here to prevent DML in loops. This test is still failing because of missing required fields. Take a look at the test errors and figure out what is missing here.